core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

```diff
@@ -504,6 +509,131 @@ class ReplicaManagerTest {
+
+    // Mock network client to show leader offset of 5
+    val countdownLatch = new CountDownLatch(1)
+    val quota = QuotaFactory.instantiate(config, metrics, time, "")
+    val blockingSend = new ReplicaFetcherMockBlockingSend(Map(new TopicPartition(topic, 0) -> new EpochEndOffset(3, 5)).asJava, BrokerEndPoint(1, "host1" ,1), time)
+    val replicaManager = new ReplicaManager(config, metrics, time, kafkaZkClient, mockScheduler, mockLogMgr,
+      new AtomicBoolean(false), quota, mockBrokerTopicStats,
+      metadataCache, mockLogDirFailureChannel, mockProducePurgatory, mockFetchPurgatory,
+      mockDeleteRecordsPurgatory, Option(this.getClass.getName)) {
+
+      override protected def createReplicaFetcherManager(metrics: Metrics, time: Time, threadNamePrefix: Option[String], quotaManager: ReplicationQuotaManager): ReplicaFetcherManager = {
+        new ReplicaFetcherManager(config, this, metrics, time, threadNamePrefix, quotaManager) {
+
+          override def createFetcherThread(fetcherId: Int, sourceBroker: BrokerEndPoint): AbstractFetcherThread = {
+            new ReplicaFetcherThread(s"ReplicaFetcherThread-$fetcherId", fetcherId, sourceBroker, config, replicaManager, metrics, time, quota.follower, Some(blockingSend)) {
+              override def doWork() = {
+                // In case the thread starts before the partition is added by AbstractFetcherManager, add it here (it's a no-op if already added)
+                addPartitions(Map(new TopicPartition(topic, 0) -> 0L))
+                super.doWork()
+
+                // Shut the thread down after one iteration to avoid double-counting truncations
+                initiateShutdown()
+                countdownLatch.countDown()
+              }
+            }
+          }
+        }
+      }
+    }
+
+    // Initialize partition state to follower, with leader = 1, leaderEpoch = 1
+    val partition = replicaManager.getOrCreatePartition(new TopicPartition(topic, 0))
+    partition.getOrCreateReplica(0)
+    partition.makeFollower(0, new LeaderAndIsrRequest.PartitionState(0, 1, 1, aliveBrokerIds.asJava, 0, aliveBrokerIds.asJava, false), 0)
+
+    // Make local partition a follower - because epoch increased by more than 1, truncation should trigger even though leader does not change
+    val leaderAndIsrRequest0 = new LeaderAndIsrRequest.Builder(ApiKeys.LEADER_AND_ISR.latestVersion, 0, 0,
+      collection.immutable.Map(new TopicPartition(topic, 0) -> new LeaderAndIsrRequest.PartitionState(0, 1, 3, aliveBrokerIds.asJava, 0, aliveBrokerIds.asJava, false)).asJava,
+      Set(new Node(0, "host1", 0), new Node(1, "host2", 1)).asJava).build()
+    replicaManager.becomeLeaderOrFollower(1, leaderAndIsrRequest0, (_, followers) => assertEquals(0, followers.head.partitionId))
+    assertTrue(countdownLatch.await(1000L, TimeUnit.MILLISECONDS))
+
+    // Make local partition a follower - because epoch increased by only 1 and leader did not change, truncation should not trigger
```

Comment:
In my humble opinion (and from what I've read in TDD books), this should two separate tests. Splitting and naming the tests more explicitly would help the developer be more clear on what has broken if the test fails.
So if `testBecomeFollowerWhenLeaderIsUnchanged` fails, you don't know what's wrong exactly. 
Where as if `testDoesNotTruncateLogOnUnchangedLeaderAndSameOrCloseEpoch` or `testTruncatesLogOnUnchangedLeaderAndFarOffEpoch` fails, you immediately know what has caused the failure.
Another thing is that these explicit method names serve as good documentation. It's always great when you can open a test class and realize what it is and is not testing by simply reading the method names
I'd like to hear other's thoughts on this too

Reply:
Yeah, I tend to agree, though I'm mostly satisfied if the test cases exist at all. But yeah, more focused test cases tend to be smaller and easier to understand.